File: samples/classification/mobilevit/mobilevit.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/mobile_vit.hpp"
#include "common/utils.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "mobilevit_s-fp16-none-1_3_224_224-vacc/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/mobilevit_rgbplanar.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("label_file", '\0', "label file", false,
                        "../data/labels/imagenet.txt");
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/cat.jpg");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

std::vector<int> Argsort(const vsx::Tensor& array) {
  const int array_len = array.GetSize();
  const float* array_data = array.Data<float>();
  std::vector<int> array_index(array_len);
  for (int i = 0; i < array_len; ++i) array_index[i] = i;

  std::sort(array_index.begin(), array_index.end(), [=](int pos1, int pos2) {
    return (array_data[pos1] > array_data[pos2]);
  });

  return array_index;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  auto labels = vsx::LoadLabels(args.get<std::string>("label_file"));
  const int batch_size = 1;
  vsx::MobileVit classifier(args.get<std::string>("model_prefix"),
                            args.get<std::string>("vdsp_params"), batch_size,
                            args.get<uint32_t>("device_id"));
  auto image_format = classifier.GetFusionOpIimageFormat();

  if (args.get<std::string>("dataset_filelist").empty()) {
    vsx::Image image;
    vsx::MakeVsxImage(args.get<std::string>("input_file"), image, image_format);
    auto result = classifier.Process(image);
    auto index = Argsort(result);
    const float* array_data = result.Data<float>();
    std::cout << "Top5:\n";
    for (int i = 0; i < 5; i++) {
      std::cout << i << "th, score: " << std::setprecision(4)
                << array_data[index[i]] << ", class name: " << labels[index[i]]
                << std::endl;
    }
  } else {
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    auto dataset_root = args.get<std::string>("dataset_root");
    std::ofstream outfile(args.get<std::string>("dataset_output_file"));
    CHECK(outfile.is_open())
        << "Failed to open: " << args.get<std::string>("dataset_output_file");
    for (auto file : filelist) {
      auto fullname = file;
      if (!dataset_root.empty()) fullname = dataset_root + "/" + file;
      std::cout << fullname << std::endl;
      vsx::Image image;
      vsx::MakeVsxImage(fullname, image, image_format);
      auto result = classifier.Process(image);
      auto index = Argsort(result);
      const float* array_data = result.Data<float>();

      for (int i = 0; i < 5; i++) {
        outfile << file << ": "
                << "top-" << i << " id: " << index[i]
                << ", prob: " << std::setprecision(8) << array_data[index[i]]
                << ", class name: " << labels[index[i]] << std::endl;
      }
    }
    outfile.close();
  }

  return 0;
}

File: samples/classification/mobilevit/mobilevit.py
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)

from common.mobile_vit import MobileVit
import common.utils as utils

import numpy as np
import cv2
import argparse


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/mobilevit_s-fp16-none-1_3_224_224-vacc/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--vdsp_params",
        default="../../../../data/configs/mobilevit_rgbplanar.json",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="../../../../data/images/cat.jpg",
        help="input file",
    )
    parser.add_argument(
        "--label_file",
        default="../../../../data/labels/imagenet.txt",
        help="label file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="dataset filelst",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        help="dataset output file",
    )
    args = parser.parse_args()
    return args


def load_labels(file):
    with open(file) as fin:
        return [line.strip() for line in fin.readlines()]


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    labels = load_labels(args.label_file)

    model = MobileVit(
        args.model_prefix,
        args.vdsp_params,
        batch_size=batch_size,
        device_id=args.device_id,
    )
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Failed to read input file: {args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        result = model.process(vsx_image)
        index = np.argsort(result)[0][::-1]
        print("Top5:")
        for i in range(5):
            print(
                f"{i}th, score: {result[0, index[i]]:.4f}, class name: {labels[index[i]]}"
            )
    else:
        with open(args.dataset_filelist, "rt") as f:
            filelist = [line.strip() for line in f.readlines()]
        with open(args.dataset_output_file, "wt") as fout:
            for file in filelist:
                fullname = os.path.join(args.dataset_root, file)
                print(fullname)
                cv_image = cv2.imread(fullname)
                assert cv_image is not None, f"Failed to read input file: {fullname}"
                vsx_image = utils.cv_bgr888_to_vsximage(
                    cv_image, image_format, args.device_id
                )
                result = model.process(vsx_image)
                index = np.argsort(result)[0][::-1]
                for i in range(5):
                    fout.write(
                        f"{file}: top-{i} id: {index[i]}, prob: {result[0, index[i]]}, class name: {labels[index[i]]}\n"
                    )
